projects/igniteui-angular/src/lib/services/exporter-common/base-export-service.ts

```diff
@@ -318,15 +325,174 @@ export abstract class IgxBaseExporter {
+    private prepareIslandData(islandGrid: IgxHierarchicalGridComponent, data: any): any {
+        if (islandGrid !== undefined) {
+            const hasFiltering = (islandGrid.filteringExpressionsTree &&
+                    islandGrid.filteringExpressionsTree.filteringOperands.length > 0) ||
+                (islandGrid.advancedFilteringExpressionsTree &&
+                    islandGrid.advancedFilteringExpressionsTree.filteringOperands.length > 0);
+
+            const hasSorting = islandGrid.sortingExpressions &&
+                islandGrid.sortingExpressions.length > 0;
+
+            const skipOperations =
+                (!hasFiltering || !this.options.ignoreFiltering) &&
+                (!hasSorting || !this.options.ignoreSorting);
+
+                if (skipOperations) {
+                    data = islandGrid.filteredSortedData;
+                } else {
+                    if (hasFiltering && !this.options.ignoreFiltering) {
+                        const filteringState: IFilteringState = {
+                            expressionsTree: islandGrid.filteringExpressionsTree,
+                            advancedExpressionsTree: islandGrid.advancedFilteringExpressionsTree,
+                        };
+
+                        filteringState.strategy = islandGrid.filterStrategy;
+
+                        data = DataUtil.filter(data, filteringState, islandGrid);
+                    }
+
+                    if (hasSorting && !this.options.ignoreSorting) {
+                        this._sort = cloneValue(islandGrid.sortingExpressions[0]);
+
+                        data = DataUtil.treeGridSort(data, islandGrid.sortingExpressions, islandGrid.sortStrategy);
+                    }
+                }
+        }
+
+        return data;
+    }
+
+    private getAllChildColumnsAndData(island: IgxRowIslandComponent,
+        childData: any, expansionStateVal: boolean, childGrids: any, grid: any) {
+        const islandColumnList = island.childColumns.toArray();
+        const modifiedColumns = this.getColumns(islandColumnList);
+
+        const headerRecord: IExportRecord = {
+            data: modifiedColumns.colList,
+            level: island.level,
+            type: ExportRecordType.HeaderRecord,
+            owner: island,
+            hidden: !expansionStateVal
+        };
+
+        if (childData && childData.length > 0) {
+            this.flatRecords.push(headerRecord);
+
+            for (const rec of childData) {
+                const exportRecord: IExportRecord = {
+                    data: rec,
+                    level: island.level,
+                    type: ExportRecordType.HierarchicalGridRecord,
+                    owner: island,
+                    hidden: !expansionStateVal
+                };
+
+                this.flatRecords.push(exportRecord);
+
+                if (island.children.length > 0) {
+                    const islandGrid = childGrids.filter(g => g.key === island.key && g.data.some(d => d === rec))[0];
```

Comment:
This doesn't seem to be the best approach for getting the correct grid:
`g.data.some(d => d === rec)`.